torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -63,197 +64,154 @@ static int64_t count_specified_dimensions(PyObject* index) {
+static inline Variable sequenceToVariable(c10::DispatchKey dispatch_key, PyObject* seq) {
   return torch::utils::indexing_tensor_from_data(dispatch_key, kLong, c10::nullopt, seq);
 }
 
-static Variable valueToTensor(c10::TensorOptions options, PyObject* value) {
+static inline Variable valueToTensor(c10::TensorOptions options, PyObject* value, const at::Device& device) {
   if (THPVariable_Check(value)) {
     return reinterpret_cast<THPVariable*>(value)->cdata;
   }
   at::AutoNonVariableTypeMode guard;
   if (THPUtils_checkLong(value) || PyBool_Check(value)) {
-    return at::scalar_tensor(Scalar(THPUtils_unpackLong(value)), options);
+    return at::indexing::scalarToTensor(Scalar(THPUtils_unpackLong(value)), options, device);
   }
   if (PyFloat_Check(value)) {
-    return at::scalar_tensor(Scalar(THPUtils_unpackDouble(value)), options);
+    return at::indexing::scalarToTensor(Scalar(THPUtils_unpackDouble(value)), options, device);
   }
   throw TypeError(
     "can't assign a %s to a %s",
     Py_TYPE(value)->tp_name,
     torch::utils::options_to_string(options).c_str());
 }
 
-static Variable boolToIndexingTensor(const Variable& self, bool value) {
-  // booleans add a dimension of size 1. true indexes this dimension as if 0:, false as empty.
-  if (value) {
-    return at::zeros({1}, self.options().dtype(kLong));
-  } else {
-    return at::empty({0}, self.options().dtype(kLong));
+static inline void extractTensorsFromSlice(
+  PyObject* obj,
+  Tensor* start_tensor_ptr,
+  Tensor* stop_tensor_ptr,
+  Tensor* step_tensor_ptr) {
+  PySliceObject* sliceobj = (PySliceObject*)obj;
+  if (THPVariable_Check(sliceobj->start)) {
+    (*start_tensor_ptr) = THPVariable_Unpack(sliceobj->start);
+  }
+  if (THPVariable_Check(sliceobj->stop)) {
+    (*stop_tensor_ptr) = THPVariable_Unpack(sliceobj->stop);
+  }
+  if (THPVariable_Check(sliceobj->step)) {
+    (*step_tensor_ptr) = THPVariable_Unpack(sliceobj->step);
+  }
+}
+
+static inline void recordSliceTrace(const Tensor& start_tensor, const Tensor& stop_tensor, const Tensor& step_tensor) {
+  if (start_tensor.defined()) {
+    torch::jit::tracer::ArgumentStash::stashValue(std::string("start"), 1, start_tensor, torch::jit::IntType::get());
+  }
+  if (stop_tensor.defined()) {
+    torch::jit::tracer::ArgumentStash::stashValue(std::string("end"), 1, stop_tensor, torch::jit::IntType::get());
+  }
+  if (step_tensor.defined()) {
+    torch::jit::tracer::ArgumentStash::stashValue(std::string("step"), 1, step_tensor, torch::jit::IntType::get());
   }
 }
 
-static Variable applySlicing(const Variable& self, PyObject* index, variable_list& outIndices) {
-  int64_t size = PyTuple_GET_SIZE(index); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
+static inline void recordSelectTrace(const Tensor& index_tensor) {
+  torch::jit::tracer::ArgumentStash::stashValue(std::string("index"), 1, index_tensor, torch::jit::IntType::get());
+}
+
+static inline PyObject* convertToPythonInt(PyObject* obj) {
```

Comment:
This is use after free. The old code used THPObjectPtr to manage the lifetime of the newly allocated `idx`, but here you are returning a `PyObject*` from the owning `idx`, which then goes out of scope after this function goes out of scope. This is **really really bad**.

Reply:
Honestly, I wouldn't have factored this into a separate function at all.